File: raknet-c/libcat.cpp
#include <RakNet/NativeFeatureIncludes.h>

#if LIBCAT_SECURITY == 1
#include <RakNet/SecureHandshake.h>

using namespace cat;

#include "libcat.h"

CookieJar *CookieJar_new()
{
    return new CookieJar();
}
void CookieJar_drop(CookieJar *self)
{
    delete self;
}

AuthenticatedEncryption *AuthenticatedEncryption_new()
{
    return new AuthenticatedEncryption();
}

bool AuthenticatedEncryption_GenerateProof(AuthenticatedEncryption *self, unsigned char *local_proof, int proof_bytes)
{
    return self->GenerateProof(local_proof, proof_bytes);
}

bool AuthenticatedEncryption_ValidateProof(AuthenticatedEncryption *self, const unsigned char *remote_proof, int proof_bytes)
{
    return self->ValidateProof(remote_proof, proof_bytes);
}

bool AuthenticatedEncryption_Decrypt(AuthenticatedEncryption *self, unsigned char *buffer, unsigned int *buf_bytes)
{
    return self->Decrypt(buffer, *buf_bytes);
}
bool AuthenticatedEncryption_Encrypt(AuthenticatedEncryption *self, unsigned char *buffer, unsigned int buffer_bytes, unsigned int *msg_bytes)
{
    return self->Encrypt(buffer, buffer_bytes, *msg_bytes);
}
void AuthenticatedEncryption_drop(AuthenticatedEncryption *self)
{
    delete self;
}

ClientEasyHandshake *ClientEasyHandshake_new()
{
    return new ClientEasyHandshake();
}
bool ClientEasyHandshake_GenerateServerKey(ClientEasyHandshake *self, void *out_public_key, void *out_private_key)
{
    return self->GenerateServerKey(out_public_key, out_private_key);
}
bool ClientEasyHandshake_Initialize(ClientEasyHandshake *self, const void *in_public_key)
{
    return self->Initialize(in_public_key);
}

bool ClientEasyHandshake_SetIdentity(ClientEasyHandshake *self, const void *in_public_key,
                                     const void *in_private_key)
{
    return self->SetIdentity(in_public_key, in_private_key);
}

bool ClientEasyHandshake_GenerateChallenge(ClientEasyHandshake *self, void *out_challenge)
{
    return self->GenerateChallenge(out_challenge);
}

bool ClientEasyHandshake_ProcessAnswer(ClientEasyHandshake *self, const void *in_answer,
                                       AuthenticatedEncryption *auth_enc)
{
    return self->ProcessAnswer(in_answer, auth_enc);
}

bool ClientEasyHandshake_ProcessAnswerWithIdentity(ClientEasyHandshake *self, const void *in_answer,
                                                   void *out_identity,
                                                   AuthenticatedEncryption *auth_enc)
{
    return self->ProcessAnswerWithIdentity(in_answer, out_identity, auth_enc);
}
void ClientEasyHandshake_drop(ClientEasyHandshake *self)
{
    delete self;
}

ServerEasyHandshake *ServerEasyHandshake_new()
{
    return new ServerEasyHandshake();
}
bool ServerEasyHandshake_GenerateServerKey(ServerEasyHandshake *self, void *out_public_key, void *out_private_key)
{
    return self->GenerateServerKey(out_public_key, out_private_key);
}
void ServerEasyHandshake_FillCookieJar(ServerEasyHandshake *self, CookieJar *jar)
{
    self->FillCookieJar(jar);
}
bool ServerEasyHandshake_Initialize(ServerEasyHandshake *self, const void *in_public_key,
                                    const void *in_private_key)
{
    return self->Initialize(in_public_key, in_private_key);
}
bool ServerEasyHandshake_ProcessChallenge(ServerEasyHandshake *self, const void *in_challenge,
                                          void *out_answer,
                                          AuthenticatedEncryption *auth_enc)
{
    return self->ProcessChallenge(in_challenge, out_answer, auth_enc);
}
bool ServerEasyHandshake_VerifyInitiatorIdentity(ServerEasyHandshake *self, const void *in_answer,
                                                 const void *in_proof,
                                                 void *out_public_key)
{
    return self->VerifyInitiatorIdentity(in_answer, in_proof, out_public_key);
}
void ServerEasyHandshake_drop(ServerEasyHandshake *self)
{
    delete self;
}
#endif

File: src/libcat.rs
use std::io::Write;

use crate::ffi;

pub const PRIVATE_KEY_BYTES: usize = ffi::PRIVATE_KEY_BYTES as usize;
pub const PUBLIC_KEY_BYTES: usize = ffi::PUBLIC_KEY_BYTES as usize;
pub const CHALLENGE_BYTES: usize = ffi::CHALLENGE_BYTES as usize;
pub const ANSWER_BYTES: usize = ffi::ANSWER_BYTES as usize;

pub const OVERHEAD_BYTES: usize = ffi::OVERHEAD_BYTES as usize;

pub struct CookieJar {
    inner: *mut ffi::CookieJar,
}
unsafe impl Sync for CookieJar {}
unsafe impl Send for CookieJar {}

impl CookieJar {
    pub fn new() -> Self {
        Self {
            inner: unsafe { ffi::CookieJar_new() },
        }
    }
}

impl Drop for CookieJar {
    fn drop(&mut self) {
        unsafe { ffi::CookieJar_drop(self.inner) }
    }
}

pub struct AuthenticatedEncryption {
    inner: *mut ffi::AuthenticatedEncryption,
}
unsafe impl Sync for AuthenticatedEncryption {}
unsafe impl Send for AuthenticatedEncryption {}

impl AuthenticatedEncryption {
    pub fn new() -> Self {
        Self {
            inner: unsafe { ffi::AuthenticatedEncryption_new() },
        }
    }

    pub fn GenerateProof(&mut self, length: usize) -> Option<Vec<u8>> {
        let mut v: Vec<u8> = vec![0; length];

        let res = unsafe {
            ffi::AuthenticatedEncryption_GenerateProof(self.inner, v.as_mut_ptr(), v.len() as i32)
        };

        if res {
            Some(v)
        } else {
            None
        }
    }

    pub fn ValidateProof(&mut self, proof: &[u8]) -> bool {
        unsafe {
            ffi::AuthenticatedEncryption_ValidateProof(
                self.inner,
                proof.as_ptr(),
                proof.len() as i32,
            )
        }
    }

    pub fn Encrypt(&mut self, msg: &[u8]) -> Option<Vec<u8>> {
        let mut v: Vec<u8> = Vec::with_capacity(msg.len() + OVERHEAD_BYTES);
        v.write(msg).unwrap();
        v.write(&[0; OVERHEAD_BYTES]).unwrap();

        let mut msg_bytes = msg.len() as u32;

        let res = unsafe {
            ffi::AuthenticatedEncryption_Encrypt(
                self.inner,
                v.as_mut_ptr(),
                v.len() as u32,
                &mut msg_bytes,
            )
        };

        if res {
            v.resize(msg_bytes as usize, 0);
            Some(v)
        } else {
            None
        }
    }

    pub fn Decrypt(&mut self, msg: &[u8]) -> Option<Vec<u8>> {
        let mut v = msg.to_vec();
        let mut msg_bytes = v.len() as u32;

        let res = unsafe {
            ffi::AuthenticatedEncryption_Decrypt(self.inner, v.as_mut_ptr(), &mut msg_bytes)
        };

        if res {
            v.resize(msg_bytes as usize, 0);
            Some(v)
        } else {
            None
        }
    }
}

impl Drop for AuthenticatedEncryption {
    fn drop(&mut self) {
        unsafe { ffi::AuthenticatedEncryption_drop(self.inner) }
    }
}

pub struct ClientEasyHandshake {
    inner: *mut ffi::ClientEasyHandshake,
}
unsafe impl Sync for ClientEasyHandshake {}
unsafe impl Send for ClientEasyHandshake {}

impl ClientEasyHandshake {
    pub fn new() -> Self {
        Self {
            inner: unsafe { ffi::ClientEasyHandshake_new() },
        }
    }

    pub fn GenerateServerKey(
        &mut self,
    ) -> Option<([u8; PUBLIC_KEY_BYTES], [u8; PRIVATE_KEY_BYTES])> {
        let mut public_key = [0; PUBLIC_KEY_BYTES];
        let mut private_key = [0; PRIVATE_KEY_BYTES];

        let res = unsafe {
            ffi::ClientEasyHandshake_GenerateServerKey(
                self.inner,
                public_key.as_mut_ptr() as _,
                private_key.as_mut_ptr() as _,
            )
        };

        if res {
            Some((public_key, private_key))
        } else {
            None
        }
    }

    pub fn Initialize(&mut self, public_key: &[u8; PUBLIC_KEY_BYTES]) -> bool {
        unsafe { ffi::ClientEasyHandshake_Initialize(self.inner, public_key.as_ptr() as _) }
    }

    pub fn SetIdentity(
        &mut self,
        public_key: &[u8; PUBLIC_KEY_BYTES],
        private_key: &[u8; PRIVATE_KEY_BYTES],
    ) -> bool {
        unsafe {
            ffi::ClientEasyHandshake_SetIdentity(
                self.inner,
                public_key.as_ptr() as _,
                private_key.as_ptr() as _,
            )
        }
    }

    pub fn GenerateChallenge(&mut self) -> Option<[u8; CHALLENGE_BYTES]> {
        let mut challenge: [u8; CHALLENGE_BYTES] = [0; CHALLENGE_BYTES];
        let res = unsafe {
            ffi::ClientEasyHandshake_GenerateChallenge(
                self.inner,
                (&mut challenge).as_mut_ptr() as _,
            )
        };

        if res {
            Some(challenge)
        } else {
            None
        }
    }

    pub fn ProcessAnswer(
        &mut self,
        answer: &[u8; ANSWER_BYTES],
    ) -> Option<AuthenticatedEncryption> {
        let authenticated_encryption = AuthenticatedEncryption::new();
        let res = unsafe {
            ffi::ClientEasyHandshake_ProcessAnswer(
                self.inner,
                answer.as_ptr() as _,
                authenticated_encryption.inner,
            )
        };

        if res {
            Some(authenticated_encryption)
        } else {
            None
        }
    }
}

impl Drop for ClientEasyHandshake {
    fn drop(&mut self) {
        unsafe { ffi::ClientEasyHandshake_drop(self.inner) }
    }
}

pub struct ServerEasyHandshake {
    _cookie_jar: CookieJar,
    inner: *mut ffi::ServerEasyHandshake,
}
unsafe impl Sync for ServerEasyHandshake {}
unsafe impl Send for ServerEasyHandshake {}

impl ServerEasyHandshake {
    pub fn new() -> Self {
        let inner = unsafe { ffi::ServerEasyHandshake_new() };
        let cookie_jar = CookieJar::new();

        unsafe {
            ffi::ServerEasyHandshake_FillCookieJar(inner, cookie_jar.inner);
        }

        Self {
            _cookie_jar: cookie_jar,
            inner,
        }
    }

    pub fn GenerateServerKey(
        &mut self,
    ) -> Option<([u8; PUBLIC_KEY_BYTES], [u8; PRIVATE_KEY_BYTES])> {
        let mut public_key = [0; PUBLIC_KEY_BYTES];
        let mut private_key = [0; PRIVATE_KEY_BYTES];

        let res = unsafe {
            ffi::ServerEasyHandshake_GenerateServerKey(
                self.inner,
                public_key.as_mut_ptr() as _,
                private_key.as_mut_ptr() as _,
            )
        };

        if res {
            Some((public_key, private_key))
        } else {
            None
        }
    }

    pub fn Initialize(
        &mut self,
        public_key: &[u8; PUBLIC_KEY_BYTES],
        private_key: &[u8; PRIVATE_KEY_BYTES],
    ) -> bool {
        unsafe {
            ffi::ServerEasyHandshake_Initialize(
                self.inner,
                public_key.as_ptr() as _,
                private_key.as_ptr() as _,
            )
        }
    }

    pub fn ProcessChallenge(
        &mut self,
        challenge: &[u8; CHALLENGE_BYTES],
    ) -> Option<(AuthenticatedEncryption, [u8; ANSWER_BYTES])> {
        let mut answer: [u8; ANSWER_BYTES] = [0; ANSWER_BYTES];
        let authenticated_encryption = AuthenticatedEncryption::new();

        let res = unsafe {
            ffi::ServerEasyHandshake_ProcessChallenge(
                self.inner,
                challenge.as_ptr() as _,
                answer.as_mut_ptr() as _,
                authenticated_encryption.inner,
            )
        };

        if res {
            Some((authenticated_encryption, answer))
        } else {
            None
        }
    }
}

impl Drop for ServerEasyHandshake {
    fn drop(&mut self) {
        unsafe { ffi::ServerEasyHandshake_drop(self.inner) }
    }
}
